ui/dashboards/src/templating/core.tsx

```diff
@@ -0,0 +1,269 @@
+// Copyright 2022 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { createContext, useContext } from 'react';
+import { createStore, useStore } from 'zustand';
+import { immer } from 'zustand/middleware/immer';
+import { devtools } from 'zustand/middleware';
+import { TemplateVariableContext } from '@perses-dev/plugin-system';
+import {
+  VariableDefinitions,
+  VariablesState,
+  VariableState,
+  VariableName,
+  VariableValue,
+  ListVariableDefinition,
+  VariableDefinition,
+  DEFAULT_ALL_VALUE as ALL_VALUE,
+} from '@perses-dev/core';
+
+type TemplateVariableSrv = {
+  variableDefinitions: VariableDefinitions;
+  variableState: VariablesState;
+  setVariableValue: (variableName: VariableName, value: VariableValue) => void;
+  loadTemplateVariable: (name: VariableName) => Promise<void>;
+};
+
+const TemplateVariableSrvContext = createContext<ReturnType<typeof createTemplateVariableSrvStore> | undefined>(
+  undefined
+);
+function useTemplateVariableSrvContext() {
+  const context = useContext(TemplateVariableSrvContext);
+  if (!context) {
+    throw new Error('TemplateVariableSrvContext not initialized');
+  }
+  return context;
+}
+
+export function useTemplateVariableValues(variableNames?: string[]) {
+  const store = useTemplateVariableSrvContext();
+  const state = useStore(
+    store,
+    (s) => {
+      const names = variableNames ?? Object.keys(s.variableState);
+      const vars: VariablesState = {};
+      names.forEach((name) => {
+        const varState = s.variableState[name];
+        if (!varState) {
+          return;
+        }
+        vars[name] = varState;
+      });
+      return vars;
+    },
+    (left, right) => {
+      return JSON.stringify(left) === JSON.stringify(right);
+    }
```

Comment:
That's interesting. I assume this is some kind of compare function? What's it used for? (Just for my own curiosity.)

Reply:
This was kind of a left over artifact of testing zustand selectors and how they worked. Basically, return the same state if the template variables that are selected didn't change. I'll probably clean this up